LUCENE-7603: Support Graph Token Streams in QueryBuilder

mattweber opened the pull request:
Adds support for handling graph token streams inside the
QueryBuilder util class used by query parsers.

mattweber commented:
Rebased against master, added missing ASF header.

dsmiley commented on lucene/core/src/java/org/apache/lucene/util/graph/GraphTokenStreamFiniteStrings.java:
```diff
@@ -0,0 +1,294 @@
+      int posInc = posIncAtt.getPositionIncrement();
+      assert pos > -1 || posInc > 0;
+
+      if (posInc > 1) {
```
This seems like a notable limitation that should be documented in javadocs somewhere.  Can't we support holes without demanding the stream use '*' ?  And might there be a test for this?

mattweber replied:
@dsmiley That was actually pulled out of the existing [TokenStreamToTermAutomatonQuery.java](https://github.com/apache/lucene-solr/blob/master/lucene/sandbox/src/java/org/apache/lucene/search/TokenStreamToTermAutomatonQuery.java#L77).  Let me look into it more.

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/graph/GraphTokenStreamFiniteStrings.java:
```diff
@@ -114,21 +127,20 @@ private void build(final TokenStream in) throws IOException {
-        throw new IllegalArgumentException("cannot handle holes; to accept any term, use '*' term");
-      }
+      // always use inc 1 while building, but save original increment
+      int fakePosInc = posInc > 1 ? 1 : posInc;
```
Maybe just `Math.min(1, posInc)` instead?

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/graph/GraphTokenStreamFiniteStrings.java:
```diff
@@ -210,82 +215,20 @@ private void finish() {
+  private int getTermID(int incr, BytesRef term) {
     Integer id = termToID.get(term);
-    if (id == null) {
+    if (incr > 1 || id == null) {
```
Hmm doesn't this mean that if the same term shows up, but with different `incr`, that it will get different `id` assigned?  But I think that is actually fine, since nowhere here do we depend on / expect that the same term must have the same id.

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/graph/GraphTokenStreamFiniteStrings.java:
```diff
@@ -174,25 +191,13 @@ private void setAccept(int state, boolean accept) {
-   */
-  private void addTransition(int source, int dest, BytesRef term) {
+  private int addTransition(int source, int dest, int incr, BytesRef term) {
     if (term == null) {
```
This can become an assert?

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/graph/GraphTokenStreamFiniteStrings.java:
```diff
@@ -0,0 +1,237 @@
+  /**
+   * Gets the list of finite string token streams from the given input graph token stream.
+   */
+  public List<TokenStream> getTokenStreams(final TokenStream in) throws IOException {
```
Could we make this method private, make this class's constructor private, and add a `static` method here, the sole public method on this class, that receives the incoming `TokenStream` and returns the resulting `TokenStream[]`?  Otherwise the API is sort of awkard, since e.g. this method seems like a getter yet it's doing lots of side-effects under the hood ...

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/graph/GraphTokenStreamFiniteStrings.java:
```diff
@@ -0,0 +1,237 @@
+      // always use inc 1 while building, but save original increment
+      int fakePosInc = posInc > 1 ? 1 : posInc;
+
+      assert pos > -1 || fakePosInc > 0;
```
Can we upgrade this to a real `if`?  I.e. we need a well-formed `TokenStream` input ... it cannot have `posInc=0` on its first token.

mattweber commented:
@mikemccand I addressed you comments.  I also added some more tests and fixed a bug that would yield wrong increment when a term that had previously been seen was found again with an increment of 0.  Tests were added.  I have squashed these changes with the previous commit so it is clear to see the difference between the original PR which did not support position increments and the new one that does.

dsmiley commented on lucene/core/src/java/org/apache/lucene/util/graph/GraphTokenStreamFiniteStrings.java:
```diff
@@ -210,85 +199,41 @@ private void finish() {
+    assert term != null;
+    boolean isStackedGap = incr == 0 && prevIncr > 1;
+    boolean hasGap = incr > 1;
+    term = BytesRef.deepCopyOf(term);
```
The deepCopyOf is only needed if you generate a new ID, not for an existing one.  

BTW... have you seen BytesRefHash?  I think re-using that could minimize the code here to deal with this stuff.

dsmiley commented on lucene/core/src/java/org/apache/lucene/util/graph/GraphTokenStreamFiniteStrings.java:
```diff
@@ -210,85 +199,41 @@ private void finish() {
    */
   private void finish(int maxDeterminizedStates) {
     Automaton automaton = builder.finish();
-
```
So all this code here removed wasn't needed after all?  It's nice to see it all go away (less to maintain / less complexity) :-)

dsmiley commented on lucene/core/src/java/org/apache/lucene/util/graph/GraphTokenStreamFiniteStrings.java:
```diff
@@ -80,22 +77,41 @@ public boolean incrementToken() throws IOException {
     List<TokenStream> tokenStreams = new ArrayList<>();
     final FiniteStringsIterator finiteStrings = new FiniteStringsIterator(det);
     for (IntsRef string; (string = finiteStrings.next()) != null; ) {
       final BytesRef[] tokens = new BytesRef[string.length];
```
Hmm; rather than materializing an array of tokens and increments, maybe you could simply give the IntsRefString  to BytesRefArrayTokenStream (and make BRATS not static) so that it could do this on the fly?  Not a big deal either way (current or my proposal).  If you do as I suggest then BRATS would no longer be a suitable name; maybe simply FiniteStringTokenStream or CustomTokenStream.

mattweber commented:
Thanks @dsmiley!  I have just pushed up code with your suggestions except for using `BytesRefHash` due to the fact we might have the same `BytesRef` but need a different id because we have position gap.

This has been great, love the feedback!

mikemccand commented:
This change looks great to me!  What an awesome improvement, to properly use graph token streams at search time so multi-token synonyms are correct.

I'll push this in a few days once I'm back home unless someone pushes first (@dsmiley feel free)...

Thank you @mattweber!

mikemccand commented:
I've merged this into Lucene's master (7.0), and I'm working on 6.x (#130) now.  Thanks @mattweber! Can you close this?
